bugzoo/mgr/coverage/extractor.py

```diff
@@ -9,265 +9,96 @@
+    provided set of instructions.
+    """
+    @classmethod
+    def register(cls, name: str) -> None:
+        logger.info("registering coverage extractor [%s] under name [%s]",
+                    cls, name)
+
+        if name in _NAME_TO_EXTRACTOR:
+            raise exceptions.NameInUseError(name)
+        if cls in _EXTRACTOR_TO_NAME:
+            m = "coverage extractor [{}] is already registered under name: {}"
+            m = m.format(cls, _EXTRACTOR_TO_NAME[cls])
+            raise Exception(m)
 
-        # copy the contents of the temporary file to the host machine
-        (_, fn_temp_host) = tempfile.mkstemp(suffix='.bugzoo')
         try:
-            mgr_ctr.copy_from(container, fn_temp_ctr, fn_temp_host)
-            with open(fn_temp_host, 'r') as fh:
-                report = fh.read()
-        finally:
-            os.remove(fn_temp_host)
+            instructions = cls.Instructions
+        except UnboundLocalError:
+            m = "coverage extractor doesn't provide an 'Instructions' class"
+            raise Exception(m)
+        if not issubclass(instructions, CoverageInstructions):
+            m = "'Instructions' class should subclass 'CoverageInstructions'"
+            raise Exception(m)
+
+        instructions.register(name)
+        _NAME_TO_EXTRACTOR[name] = cls
+        _EXTRACTOR_TO_NAME[cls] = name
+
+        logger.info("registered coverage extractor [%s] under name [%s]",
+                    cls, name)
+
+    @classmethod
+    def register_as_default(cls, language: Language) -> None:
+        logger.info("registering coverage extractor [{}] as default for {}",
+                    cls, language)
+
+        if cls not in _EXTRACTOR_TO_NAME:
+            m = "coverage extractor [{}] has not been registered under a name"
+            m = m.format(cls)
+            raise Exception(m)
+
+        name = _EXTRACTOR_TO_NAME[cls]
+        cls_instructions = CoverageInstructions.find(name)
+        cls_instructions.register_as_default(language)
+        logger.info("registered coverage extractor [{}] as default for {}",
+                    cls, language)
 
-        t_start = timer()
-        logger_c.debug("Parsing gcovr XML report.")
-        res = self._parse_report(report)
-        logger_c.debug("Finished parsing gcovr XML report (took %.2f seconds).", timer() - t_start)  # noqa: pycodestyle
-        logger_c.debug("Finished extracting coverage information")
-        return res
+    @staticmethod
+    def build(installation: 'BugZoo',
+              container: Container
+              ) -> 'CoverageExtractor':
+        bug = installation.bugs[container.bug]  # type: Bug
+        instructions = bug.instructions
+        if instructions is None:
+            raise exceptions.NoCoverageInstructions
+
+        name = instructions.__class__.registered_under_name()
+        extractor_cls = _NAME_TO_EXTRACTOR[name]
+        extractor = extractor_cls(installation, container, instructions)
```

Comment:
You expect `CoverageExtractor` to take these 3 arguments in `__init__` but that's not specified anywhere. If I want to write a `CoverageExtractor` I need to look at an example to understand what's happening.

Reply:
Added instructions to `CoverageExtractor` docstring.